### PR TITLE
Rust interface

### PR DESCRIPTION
**Pull request by jeremylt**

### CODE REVIEW

**Comment by jeremylt** [comment]
Currently: 
```
cd rust
cargo build
```
builds the libCEED + Rust bindings

File: rust/src/basis.rs
Comment (jeremylt) [inline rust/src/basis.rs:29]: Do we want to just put all enums in `ceed` like was done with the Python interface? Then the user always uses `ceed::EnumType::EnumValue` like `ceed::MemType::Host`

**Comment by jeremylt** [comment]
The big ToDo is QFunctions and Operators.

The question there is what's the Rust-y way to handle our different subclasses of QFunctions/Operators. There are some common functions between our two variants of QFunctions (or Operators), but there some functions unique to each. I need to figure out the best way to express this with Rust.

QFunctions
- Add Fields
- Add Context
- Apply
QFunctionByName
- Apply

Operator
- Add Fields
- Apply
- Assemble Diag
- Multigrid Create

CompositeOperator
- Add Sub Operators
- Apply
- Assemble Diag

File: rust/src/vector.rs
Comment (jedbrown) [inline rust/src/vector.rs:321]: I'd like to get rid of the functions like `get_array`, which leak memory if the caller does not remember to `restore_array`. I don't have strong opinions about whether `VectorView` coerces (via `Deref`) to a slice or ndarray. One possible point in favor of slices is minimizing dependencies and encouraging users to use `zip` patterns for manipulating vectors in-place versus the common allocating semantics in `ndarray`.

File: rust/src/vector.rs
Comment (jedbrown) [inline rust/src/vector.rs:373]: An alternate way to do this would be to always call `CeedVectorGetArrayRead` when creating a view. I used reference counting on the Rust side (`Vector` holds a `Weak` reference while each `VectorView` holds a strong (`Rc`)) because it seems to let the compiler be a little stronger about enforcing the mutable reference rules. I'm not attached to it, however.

**Comment by jeremylt** [comment]
Unfortunately, our `const` CeedObjects like `CEED_VECTOR_ACTIVE`, etc are not `const` in the bindings but instead `static` which means we can't make a const rust `vector_active`, ect out of them. I'm not sure if this is a bindgen bug or not, but if it's not then we will have workaround in a slightly annoying way.

In Python we were able to create the analogos const objects and just call through to C
```
    def set_field(self, fieldname, restriction, basis, vector):
        """Provide a field to an Operator for use by its QFunction.
           Args:
             fieldname: name of the field (to be matched with the same name used
                          by QFunction)
             restriction: ElemRestriction
             basis: Basis in which the field resides or CEED_BASIS_COLLOCATED
                      if collocated with quadrature points
             vector: Vector to be used by Operator or CEED_VECTOR_ACTIVE
                       if field is active or CEED_VECTOR_NONE if using
                       CEED_EVAL_WEIGHT in the QFunction"""

        # libCEED call
        fieldnameAscii = ffi.new("char[]", fieldname.encode('ascii'))
        err_code = lib.CeedOperatorSetField(self._pointer[0], fieldnameAscii,
                                            restriction._pointer[0], basis._pointer[0],
                                            vector._pointer[0])
        self._ceed._check_error(err_code)
```

**Comment by jedbrown** [comment]
How did you try to implement this? Why can't we have a static object on the Rust side?

**Comment by jeremylt** [comment]
If I try to make these constants as static I get: 
```
`*mut prelude::bind_ceed::Ceed_private` cannot be shared between threads safely
```

**Comment by jeremylt** [comment]
Alright, I think that we have the CEED_VECTOR_NONE and such now

**Comment by jeremylt** [comment]
We are now in a basic functional state :tada:

I think think the biggest question now is the array interface

**Comment by jeremylt** [comment]
Updated ToDo (Feel free to edit):

- [ ] array ownership and viewing
- [ ] lots of casting between usize and i32 right now for arguments
- [ ] multigrid functions should return [op_coarse, op_prolong, op_restrict] (don't think rust likes passing in uninitalized args)
- [ ] expand documentation tests (lots of op stubs)
- [ ] ...

File: rust/src/vector.rs
Comment (jeremylt) [reply rust/src/vector.rs]: As thoroughly discussed on Slack, I dropped  `get_array` and `restore_array` and added a mutable analogue of this interface.

**Comment by jeremylt** [comment]
The current ToDos:

Impl
- [ ] Implement User QFunctions, I think we need a trampoline
- [ ] Implement QFunctionContext, need to pass around arbitrary structs

Testing
- [ ] Add examples of point block assembly for testing
- [ ] Add composite operator examples for testing

Refactor
- [ ] Refactor multigrid assembly stuff to return new Rust operator objects

**Comment by jedbrown** [comment]
I don't think contexts are needed for QFunctions written in Rust since Rust has closures and [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html) to carry that state. We can deal with GPUs later.

**Comment by jeremylt** [comment]
That will make things easier

**Comment by jeremylt** [comment]
User QFunction are now needed to proceed with the other ToDos.

I'm looking into both a way to use native Rust functions and a way to continue to use our single source files.

**Comment by jedbrown** [comment]
I think a Rust qfunction should implement a trait. There could be a trait for scalar functions and vectorized functions, with a default implementation lifting scalar to vector. There could be a `From` implementation from appropriately-typed functions so users could write their qfunction in a closure. We might be able to use a macro to provide an interface similar to what Will has done in #632.

File: rust/src/qfunction.rs
Comment (jeremylt) [inline rust/src/qfunction.rs:211]: @jedbrown @YohannDudouit 

This isn't quite right but should be close to what we need for (finally) finish this. Thoughts?

**Comment by jeremylt** [comment]
We now have support for user closures.

ToDo:
- Composite operator examples
- Documentation auto deployment

I should be done tomorrow

**Comment by jeremylt** [comment]
Documentation live link: http://ceed.exascaleproject.org/libCEED-rust-docs/ceed/index.html

**Comment by jeremylt** [comment]
Updated for the Julia PR merge. I think this should be ready to merge and then we're ready to submit to JOSS.

File: .travis.yml
Comment (jeremylt) [inline .travis.yml:154]: We'll similarly need to shift this target branch when merging

**Comment by jeremylt** [comment]
I'll update for #658 once merged

**Comment by jeremylt** [comment]
ToDo: set up doc deployment for GitHub Actions.

File: rust/src/lib.rs
Comment (jeremylt) [inline rust/src/lib.rs:591]: This is much cleaner for the users now

**Comment by jeremylt** [comment]
I think this is pretty close. Sure there are things that could be improved, but I think this is in a good functional yet experimental state.

I just need to fix the doc deployment and I think its ok to merge (and then JOSS).

File: .github/workflows/rust-documentation.yml
Comment (jeremylt) [inline .github/workflows/rust-documentation.yml:7]: Note: drop extra branch when merged

**Comment by jeremylt** [comment]
Doc deployment all set

File: rust/src/qfunction.rs
Comment (YohannDudouit) [inline rust/src/qfunction.rs:190]: I'm really not a big fan of having to copy all the pointers for each apply, I feel like we should be able to avoid it, but I don't really see how.

File: rust/src/qfunction.rs
Comment (YohannDudouit) [inline rust/src/qfunction.rs:196]: I'm wondering if we could instead directly have arrays of slices of size `MAX_QFUNCTION_FIELDS` instead of slices of slices, could that prevent having to copy all the pointers in `apply`?

File: rust/src/qfunction.rs
Comment (YohannDudouit) [inline rust/src/qfunction.rs:210]: Could we have arrays of slices here to prevent transforming `Vec` into arrays?

File: rust/src/qfunction.rs
Comment (YohannDudouit) [inline rust/src/qfunction.rs:221]: Use `[&mut [f64], MAX_QFUNCTION_FIELDS]`?

**Review by YohannDudouit — COMMENTED** [review]
Overall I think this is almost ready to be merged, I just have small suggestions/questions about the way we handle qFunctions. The qFunction code feels more complex than it should be. Otherwise very good work @jeremylt .

File: rust/src/qfunction.rs
Comment (jedbrown) [reply rust/src/qfunction.rs:221]: [arrayvec](https://docs.rs/arrayvec/0.5.2/arrayvec/struct.ArrayVec.html) gives the `Vec` interface for stack-allocated arrays, though a raw array also seems fine here.

Somewhat related to this review, I don't know how to provide interior mutability without a much more complicated type signature (e.g., using `RefCell`).
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b33c4708959797511baf72a8be6fc965

File: rust/src/qfunction.rs
Comment (jeremylt) [reply rust/src/qfunction.rs:210]: I picked vec with a runtime determined length so the user couldn't reach into the inputs/outputs array into an uninitalized entry, but I don't feel strongly about it.

File: rust/src/qfunction.rs
Comment (YohannDudouit) [reply rust/src/qfunction.rs:221]: Is it what you want?
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=112b28c2b8892bd520bc7725553eef44

Comment (jeremylt) [reply rust/src/qfunction.rs:221]: That will not allow assignment to a mutable variable, which seems desirable
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=112b28c2b8892bd520bc7725553eef44

File: rust/src/qfunction.rs
Comment (jeremylt) [reply rust/src/qfunction.rs:196]: We would still need to make slices from the raw pointers, since I'd like the users not to have to do `unsafe` things and using raw pointers is `unsafe`

File: rust/src/qfunction.rs
Comment (jeremylt) [reply rust/src/qfunction.rs:190]: I agree that the copies are ugly, but realistically, the user should never call this function (should use OperatorApply), so I don't know that the overhead here is a big deal.

File: rust/src/qfunction.rs
Comment (YohannDudouit) [reply rust/src/qfunction.rs:221]: I thought this was what @jedbrown wanted

Comment (jeremylt) [reply rust/src/qfunction.rs:221]: Well, I think its good to allow
```
mut v = &mut outputs[0];
```
and disallow
```
outputs[0] = v
```
if we can, but I feel more strongly about allowing the first.

Comment (jedbrown) [reply rust/src/qfunction.rs:221]: I wanted the interior (array entry) mutability without ownership (i.e., slice, not array). But I suppose that as a function argument, it'd be okay to move the array, as in
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=38ba82d5e1bf110ea4fe5075eabf1fe2

Trouble is that the length of the array is now part of the function signature (maybe this is also a good thing). I'm not sure the best way to express it, but it'd be better if we could just make them all function arguments. It seems like macro should be able to determine the number of arguments and mutability of those arguments in order to do the transformation, but I think we could work out that refinement in a subsequent PR.

Comment (jedbrown) [reply rust/src/qfunction.rs:221]: And the move semantics produce good code (no copies, just a bounds check).
```
example::f:
        push    rax
        cmp     rsi, 2
        jb      .LBB6_2
        add     dword ptr [rdi + 4], 50
        pop     rax
        ret

...
        mov     rdi, qword ptr [rsp + 80]
        mov     rsi, qword ptr [rsp + 88]
        call    example::f
```
for
```rust
#[inline(never)]
fn f(a: [&mut [i32]; 2]) {
    a[0][1] += 50;
}

pub fn main() {
    let mut r0 = [1, 2];
    let mut r1 = [3, 4];
    let z: [&mut [i32]; 2] = [&mut r0, &mut r1];
    z[0][1] = 5;
    println!("z {:?}", z);
    f(z);         // z moved here, but r0 and r1 are still valid
    println!("z {:?}", [r0, r1])
}
```
https://gcc.godbolt.org/z/GoKzGx

Comment (jeremylt) [reply rust/src/qfunction.rs:221]: I think when I tried this before I had to set the array length to the max possible, which has us back at 
```
[&mut [f64], MAX_QFUNCTION_FIELDS]
```

Comment (jedbrown) [reply rust/src/qfunction.rs:221]: A good use case for const generics?
https://without.boats/blog/shipping-const-generics/

File: rust/src/qfunction.rs
Comment (YohannDudouit) [reply rust/src/qfunction.rs:210]: I think what @jedbrown proposed with `ArrayVec` solves this issue.

File: rust/src/qfunction.rs
Comment (YohannDudouit) [reply rust/src/qfunction.rs:221]: I also think that we should be able to do with Rust macros (`qfunction!`) something with the same flavor as what @pazner has done for the Julia interface. DSL are also extremely easy to write with Rust which should help us write very elegantly `CeedOperator`. I agree that none of this needs to be in the first release of the Rust interface.

**Review by YohannDudouit — APPROVED** [review]

**Comment by jeremylt** [comment]
I'm using `q` here
```
let mut mass_2_comp = |
  q: usize,
  [u, qdata, ..]: [&[f64]; MAX_QFUNCTION_FIELDS],
  [v, ..]: [&mut [f64]; MAX_QFUNCTION_FIELDS],
 | -> i32
 {
   // Iterate over quadrature points
   for i in 0..q {
    v[i + 0*q] = u[i + 1*q] * qdata[i];
    v[i + 1*q] = u[i + 0*q] * qdata[i];
  }

  // Return clean error code
  0
};
```

**Comment by jeremylt** [comment]
Alrighty, we're down to
```
let mut user_f = |
  [u, weights, ..]: [&[f64]; MAX_QFUNCTION_FIELDS], // inputs
  [v, ..]: [&mut [f64]; MAX_QFUNCTION_FIELDS], // outputs
| -> i32
{
  // Iterate over quadrature points
  v
    .iter_mut()
    .zip(u.iter().zip(weights.iter()))
    .for_each(|(v, (u, w))| *v = u * w);

  // Return clean error code
  0
};
```

I think this should work for now and give us something good to work with for a future macro.

**Comment by jedbrown** [comment]
I think my push is a modest improvement.

I think it would be more idiomatic to return a `Result<(), E>` so the `?` operator can be used inside qfunctions.

**Comment by jedbrown** [comment]
I wouldn't let it block merging this branch, but perhaps we can use an iteration idiom like this.
https://github.com/lumol-org/soa-derive

**Review by jedbrown — APPROVED** [review]
I'm fine with merging this now and we can refine in future PRs. One thing that I think will be useful is to have a way to build an internal libceed.so (or libceed.a). That seems common among other packages, to avoid installation complexities that the user needs to build and set `LD_LIBRARY_PATH` separately.
https://github.com/gyscos/zstd-rs
https://github.com/blas-lapack-rs/blas-lapack-rs.github.io/wiki

File: rust/src/lib.rs
Comment (jeremylt) [inline rust/src/lib.rs:26]: @jedbrown, install instructions added.

File: rust/src/qfunction.rs
Comment (jeremylt) [inline rust/src/qfunction.rs:18]: I also added brief descriptions of each module to make the docs easier to read.

**Pull request merged**
